enhancements/api-review/custom-tags-aws.md

```diff
@@ -91,19 +90,20 @@ userTags that are specified in the infrastructure resource will merge with userT
 
 The userTags field is intended to be set at install time and is considered immutable. Components that respect this field must only ever add tags that they retrieve from this field to cloud resources, they must never remove tags from the existing underlying cloud resource even if the tags are removed from this field(despite it being immutable).
 
-If the userTags field is changed post-install, there is no guarantee about how an in-cluster operator will respond to the change. Some operators may reconcile the change and change tags on the AWS resource. Some operators may ignore the change. However, if tags are removed from userTags, the tag will not be removed from the AWS resource.
-
-### Updated design details
+In case of hosted control plane deployments, the userTags field is updated with latest updates requested by user. The merge logic gives higher precedence to userTags when there is duplicate tag found on the AWS resource.
 
-A new field `propagateUserTags` is added in GA release version. The `experimentalPropagateUserTags` field will be deprecated. In future release versions, `experimentalPropagateUserTags` will be removed.
-When both fields are set, `experimentalPropagateUserTags` takes precedence.
+If the userTags field is changed post-install, there is no guarantee about how an in-cluster operator will respond to the change. Some operators may reconcile the change and change tags on the AWS resource. Some operators may ignore the change. However, if tags are removed from userTags, the tag will not be removed from the AWS resource.
```

Comment:
You moved this paragraph, but did you mean to remove it entirely (or perhaps keep it for historical reference, but mark it as outdated)?  We've run into an obstacle with cluster-ingress-operator and cloud-provider-aws (see https://github.com/openshift/cluster-ingress-operator/pull/1148#issuecomment-2421811583), so I want to understand whether (1) handling updates and (2) handling removals in particular are hard requirements or not.

Reply:
In contrast to `cluster-ingress-operator` and `cloud-provider-aws`, I checked the behaviour with [aws-load-balancer-operator#137](https://github.com/openshift/aws-load-balancer-operator/pull/137), and it can update and remove the tags. 

```
oc get infrastructures.config.openshift.io cluster -oyaml | yq .status.platformStatus

aws:
  region: us-east-2
  resourceTags:
    - key: conflict-key1
      value: plat-value2
    - key: conflict-key2
      value: plat-value3
type: AWS

aws elbv2 describe-tags --resource-arns arn:aws:elasticloadbalancing:us-east-*******

{
    "TagDescriptions": [
        {
            "ResourceArn": "arn:aws:elasticloadbalancing:us-east-*******",
            "Tags": [
                {
                    "Key": "service.k8s.aws/resource",
                    "Value": "LoadBalancer"
                },
                {
                    "Key": "service.k8s.aws/stack",
                    "Value": "aws-load-balancer-test-default-lb-class/echoserver"
                },
                {
                    "Key": "conflict-key2",
                    "Value": "plat-value3"
                },
                {
                    "Key": "conflict-key1",
                    "Value": "plat-value2"
                },
                {
                    "Key": "elbv2.k8s.aws/cluster",
                    "Value": "ckyal-20241021-2e71f8-pmfc8"
                }
            ]
        }
    ]
}

```
Updated infrastructure status: removed `conflict-key1` and updated `conflict-key2`

```
oc get infrastructures.config.openshift.io cluster -oyaml | yq .status.platformStatus
aws:
  region: us-east-2
  resourceTags:
    - key: conflict-key2
      value: new-plat-value3
type: AWS

aws elbv2 describe-tags --resource-arns arn:aws:elasticloadbalancing:us-east-*******

{
    "TagDescriptions": [
        {
            "ResourceArn": "arn:aws:elasticloadbalancing:us-east-*******",
            "Tags": [
                {
                    "Key": "service.k8s.aws/resource",
                    "Value": "LoadBalancer"
                },
                {
                    "Key": "service.k8s.aws/stack",
                    "Value": "aws-load-balancer-test-default-lb-class/echoserver"
                },
                {
                    "Key": "conflict-key2",
                    "Value": "new-plat-value3"
                },
                {
                    "Key": "elbv2.k8s.aws/cluster",
                    "Value": "ckyal-20241021-2e71f8-pmfc8"
                }
            ]
        }
    ]
}
```

----
So, there is a bit of inconsistency between upstream [cloud-provider-aws](https://github.com/kubernetes/cloud-provider-aws) and [aws-load-balancer-controller](https://github.com/kubernetes-sigs/aws-load-balancer-controller) on their way of handling tags. 

From OpenShift (`cluster-ingress-operator` and `aws-load-balancer-operator`) side, I _think_  we can only control
- `aws-load-balancer-additional-resource-tags` annotation for `cluster-ingress-operator` , and
- `--default-tags` container arg for `aws-load-balancer-operator`

to propagate relevant tags to its consumers, which in this case are cloud-provider-aws and aws-load-balancer-controller, respectively. If we need to standardise the behaviour, then we might need to change the upstream logic.

Reply:
> You moved this paragraph, but did you mean to remove it entirely (or perhaps keep it for historical reference, but mark it as outdated)? We've run into an obstacle with cluster-ingress-operator and cloud-provider-aws (see [openshift/cluster-ingress-operator#1148 (comment)](https://github.com/openshift/cluster-ingress-operator/pull/1148#issuecomment-2421811583)), so I want to understand whether (1) handling updates and (2) handling removals in particular are hard requirements or not.

Ah!, I think it is a residue of the initial version, thanks @Miciah. In summary, all redhat supported operators (in-cluster or day2) will reconcile the tags on resources created and managed by them. Removal of tag is not supported, as it'd intervene and remove the tags un-intentionally for tags added externally by the user.

Reply:
> Removal of tag is not supported, as it'd intervene and remove the tags un-intentionally for tags added externally by the user, if any.

[aws-load-balancer-controller](https://github.com/kubernetes-sigs/aws-load-balancer-controller) provides a  command line flag called `--external-managed-tags` through which a list of tag keys can be ignored to be reconciled by the controller. 

xRef : https://kubernetes-sigs.github.io/aws-load-balancer-controller/latest/deploy/configurations/

```
external-managed-tags  stringList  AWS Tag keys that will be managed externally. Specified Tags are ignored during reconciliation
```
This means if a user wants to manage a tag externally, it has to be added in `external-managed-tags` list. If not, aws-load-balancer-controller will remove unsolicited tags, and apply tags defined in `--default-tags`. 

@alebedev87 As I can see, [aws-load-balancer-operator](https://github.com/openshift/aws-load-balancer-operator) doesn't provide a field to specify `ExternalManagedTags`. Was there any discussion around this, or it's an existing gap?